src/frequenz/sdk/microgrid/component_data.py

```diff
@@ -17,229 +19,218 @@
 from .component_states import EVChargerCableState
 
 
-class _BaseComponentData(ABC):
+@dataclass(frozen=True)
+class ComponentData(ABC):
     """A private base class for strongly typed component data classes."""
 
-    _raw: microgrid_pb.ComponentData
+    component_id: int
+    timestamp: datetime
 
-    @property
-    def component_id(self) -> int:
-        """Return the component id of the message.
+    @staticmethod
+    def from_proto(raw: microgrid_pb.ComponentData) -> ComponentData:
+        """Create ComponentData from a protobuf message.
```

Comment:
This is very weird, one comment got lost from my last review.

I said that this should probably be:

```suggestion
from typing_extensions import Self
...

    @classmethod
    def from_proto(cls: Self, raw: microgrid_pb.ComponentData) -> Self:
        """Create ComponentData from a protobuf message."""
        ...
        
        return cls(...)
```

Otherwise if we need to subclass some component without adding new attributes for some reason, the `from_proto()` will return the wrong type.

Since [`Self`](https://docs.python.org/3.11/library/typing.html#typing.Self) was introduced in Python 3.11, we need to add [`typing_extensions`](https://pypi.org/project/typing-extensions/) as a dependency to be able to use this, but I think is fine, because is a package maintained by Python itself, so it should be stable and maintained and we are not requiring Python 3.11 as the minimum version for a while, so I think it would be a good addition to the tools we can use in the SDK in terms of typing.

Reply:
Technically, it makes sense to use `@classmethod` and `cls` as the first argument, but do you really see it coming in the near future that someone will be subclassing these? I thought we wanted to agree on these data classes so that everyone sticks to them and not subclasses them.

I can add it, but somehow adding a 3rd party package just to have a type annotation in 4 places seems a bit like an overkill.

Reply:
I would add it because:

* `typing_extension` is not 3rd party, is part of the Python foundation, it's just backporting of various typing improvements available in future Python versions, so it is maintained, very limited in scope, and temporary (once we have a sufficiently new Python version as the lowest Python version supported we can drop it), is not really a library adding new features.
* I don't think it will be used only here, Python typing system is still being stabilized, and new typing tools were added on each Python version. I think `Self` is super useful in many other places, and it also provides [`@dataclass_transform()`](https://peps.python.org/pep-0681/), which it might be the *right way* to fix the typing of the `@actor` decorator (I'm not 100% sure, I just looked at it quickly).
* About using the `@staticmethod`, it's true that it will probably not happen soon that we have a problem with it, but if it happens it might be a hard-to-catch issue, that might explode in weird ways, as types won't match but all will be a subclass of `ComponentData`, so I would rather avoid it even if we don't want to add `typing_extensions` (I think having the wrong type hint is better than returning a wrong type).

Reply:
I tried, but it looks like it is supported starting from Python 3.11. `mypy` doesn't accept it either:
![image](https://user-images.githubusercontent.com/109511717/194045510-ab956394-fcfc-4ec2-8cfa-7e8812ada77c.png)